build_tools/cmake/iree_run_module_test.cmake

```diff
@@ -0,0 +1,371 @@
+#     "--function_input=f32=-10"
+#   EXPECTED_OUTPUT
+#     "f32=10"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     person_detect_int8_correctness_test
+#   MODULE_SRC
+#     "person_detect_int8.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=main"
+#     "--function_input=1x96x96x1xi8=0"
+#   EXPECTED_OUTPUT
+#     "1x2xi8=[72 -72]"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     mobilenet_v1_fp32_correctness_test
+#   MODULE_SRC
+#     "mobilenet_v1_fp32.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=main"
+#     "--function_input=1x224x224x3xf32=0"
+#   EXPECTED_OUTPUT
+#     "mobilenet_v1_fp32_expected_output.txt"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+#     "riscv32-Linux"
+# )
+
+function(iree_run_module_test)
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;MODULE_SRC;DRIVER;EXPECTED_OUTPUT;TIMEOUT"
+    "RUNNER_ARGS;LABELS;XFAIL;UNSUPPORTED_PLATFORMS;DEPS"
+    ${ARGN}
+  )
+
+  iree_get_platform(_PLATFORM)
+  if(_PLATFORM IN_LIST _RULE_UNSUPPORTED_PLATFORMS)
+    return()
+  endif()
+
+  if(NOT DEFINED _RULE_DRIVER)
+    message(SEND_ERROR "Need to specify the driver to use.")
+  endif()
+
+  set(_RUNNER_TARGET "iree-run-module")
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  file(REAL_PATH "${_RULE_MODULE_SRC}" _SRC)
+  list(APPEND _RULE_RUNNER_ARGS "--module_file=${_SRC}")
+
+  if(_RULE_EXPECTED_OUTPUT)
+    # this may be a file or a literal output. In the latter case, the
+    # extension variable will be empty.
+    cmake_path(GET _RULE_EXPECTED_OUTPUT EXTENSION LAST_ONLY _OUTPUT_FILE_TYPE)
+    if(_OUTPUT_FILE_TYPE AND NOT
+       (_OUTPUT_FILE_TYPE STREQUAL ".txt" OR _OUTPUT_FILE_TYPE STREQUAL ".npy"))
+      message(SEND_ERROR
+        "Unsupported expected output file type: ${_RULE_EXPECTED_OUTPUT}"
+      )
+    elseif(_OUTPUT_FILE_TYPE STREQUAL ".txt")
+      file(REAL_PATH "${_RULE_EXPECTED_OUTPUT}" _OUTPUT_FILE_SRC)
+      # Process the text input to remove the line breaks.
+      file(STRINGS "${_OUTPUT_FILE_SRC}" _EXPECTED_OUTPUT ENCODING UTF-8)
+      string(REPLACE ";" " " _EXPECTED_OUTPUT_STR "${_EXPECTED_OUTPUT}")
+      set(_EXPECTED_OUTPUT_STR "--expected_output=\"${_EXPECTED_OUTPUT_STR}\"")
+      list(APPEND _RULE_RUNNER_ARGS ${_EXPECTED_OUTPUT_STR})
+      list(APPEND _SRC ${_OUTPUT_FILE_SRC})
+    elseif(_OUTPUT_FILE_TYPE STREQUAL ".npy")
+      # Large npy files are not stored in the codebase. Need to download them
+      # from GCS iree-model-artifacts first and store them in the following possible
+      # paths.
+      find_file(_OUTPUT_FILE_SRC_PATH
+        NAME
+          "${_RULE_EXPECTED_OUTPUT}"
+        PATHS
+          "${CMAKE_CURRENT_SOURCE_DIR}"
+          "${CMAKE_CURRENT_BINARY_DIR}"
+          "${IREE_BENCHMARK_SUITE_DIR}"
+        NO_CACHE
```

Comment:
Do we really want `NO_CACHE` here for files too large to store in the repository? https://cmake.org/cmake/help/latest/command/find_file.html

> Warning This option should be used with caution because it can greatly increase the cost of repeated configure steps.

We could also move this outside of the build system as a pre-build step as Geoffrey has suggested for a few similar "download sources" parts of the project.

Also, with your full stack of changes, how many files are downloaded / what's their size on disk?

Reply:
It is not a download step, only to find the file from the indicated paths. If the file is not found, it will report an error in the following line. I want to ask the cmake to identify the existence of the file when the build is re-configured. 

The benchmark suite is 2GB in total, while the expected output for deeplab is ~5MB.